[REVIEW] RAII-ify `pq_chunked_state` [skip ci]

codereport opened the pull request:
Resolves https://github.com/rapidsai/cudf/issues/4735

Issue:
```cpp
  pq_chunked_state state;
  state.user_metadata = metadata;
  state.stream = stream;
  state.single_write_mode = true;
```
This code can be improved by adding a constructor and setting the members in the ctor, therefore following one of C++'s oldest best practices: RAII.

vuule commented on cpp/include/cudf/io/writers.hpp:
```diff
@@ -193,8 +193,8 @@ class writer {
    */
-  void write_chunked_begin(struct pq_chunked_state& state);                           
-  
+  auto write_chunked_begin() -> std::size_t;
```
Why did you switch to the trailing return type in this PR?

rgsl888prabhu replied:
Same concern

codereport replied:
Yea, bad habit of mine - I will switch it back. It is just my personal preference and I forget when I am refactoring or coding sometimes.

trxcllnt commented:
@codereport this is a lot of commits, does it need to be rebased over current branch-0.14?

codereport commented:
> @codereport this is a lot of commits, does it need to be rebased over current branch-0.14?

Yea, something clearly happened with my merge of `branch-0.14`. Let me figure it out.

codereport commented:
> > @codereport this is a lot of commits, does it need to be rebased over current branch-0.14?
> 
> Yea, something clearly happened with my merge of `branch-0.14`. Let me figure it out.

So I managed to fix it with a `git rebase branch-0.14`, resolving all the conflicts again and then a `git push --force`. It didn't work without the `--force` though (and I know we try to avoid those). @trxcllnt do you know a way around the `--force`?

trxcllnt commented:
@codereport yeah unfortunately `--force` is the only remedy to update a remote after rewriting the commit history via rebase. You can use the slightly safer `--force-with-lease` option if it's a collaborative branch, but both ways end up stranding any existing inline github comments.

harrism commented:
@codereport here's my reflex "PLEASE DON'T FORCE PUSH to open PR branches" message. Doing so causes the existing review comments to disassociate from the code history (you can see this above).  In general I think it's best to avoid `git rebase` except in branches that are not shared and aren't part of an open PR. 

`git merge` is almost always sufficient.

codereport commented:
> @codereport here's my reflex "PLEASE DON'T FORCE PUSH to open PR branches" message. Doing so causes the existing review comments to disassociate from the code history (you can see this above). In general I think it's best to avoid `git rebase` except in branches that are not shared and aren't part of an open PR.
> 
> `git merge` is almost always sufficient.

Hi @harrism, Not sure if you saw the back and forth between @trxcllnt and I but I was trying to avoid this at all costs and know we don't ever want to force push. I will try to avoid `rebase` in the future but IIRC at the time `merge` was not working.

harrism commented:
Sorry, you didn't say you tried to avoid rebase.  I just flag it when I see it because some contributors don't know.

codereport commented:
@vuule I have updated the PR CHANGELOG and added a more descriptive description. Please re-review PR as it has been simplified.

@OlivierNV I have revert the code that set `current_chunk_offset` outside of the ctor. It made the ctor only usable in one of the use cases and now this PR is less of an improvement but c'est la vie. Please re-review.

OlivierNV commented on cpp/src/io/parquet/chunked_state.hpp:
```diff
@@ -40,15 +43,24 @@ struct pq_chunked_state {
+  /// only used in the write_chunked() case. copied from the (optionally) user supplied
+  /// argument to write_parquet_chunked_begin()
+  bool single_write_mode;
+
```
This adds a default constructor (needed for the chunked path and/or python iirc), but also removes the default init for user_metadata and single_write_mode (@nvdbaranec should confirm, but that looks somewhat sketchy to me).

codereport replied:
I can add back the `nullptr` default for metatdata.

nvdbaranec commented on cpp/src/io/parquet/chunked_state.hpp:
```diff
@@ -40,15 +43,24 @@ struct pq_chunked_state {
+  pq_chunked_state(table_metadata const* metadata,
+                   SingleWriteMode mode = SingleWriteMode::NO,
+                   cudaStream_t str     = 0)
+    : user_metadata{metadata}, single_write_mode{mode == SingleWriteMode::YES}, stream{str}
```
From Olivier's comment above

Looks like the constructor flips the default value from false to true, which I think is backwards.   Since the default use case for this is indeed "chunked" (multi-write) it should be set that way.   The single-write flag is there so that non-chunked write can quietly reuse the chunked-write code internally.  

Summary : the constructor should set single_write_mode to SingleWriteMode::NO
```suggestion
    : user_metadata{metadata}, single_write_mode{mode == SingleWriteMode::NO}, stream{str}
```

codereport replied:
@nvdbaranec This initialization list in the ctor only gets called if invoked. And if the user passes `SingleWriteMode::Yes` then ultimately we want the boolean to be set to true. This change would make it set to `false`. If we want the default to be `false` when the default is used, we should be changed the default in the parameter list of the function definition. I will make that change.

nvdbaranec replied:
Ugh, you're right.  I read this backwards.  Sorry :)
